sklearn/utils/estimator_checks.py

```diff
@@ -1106,60 +1104,105 @@ def check_sample_weights_invariance(name, estimator_orig, kind="ones"):
         ],
         dtype=np.float64,
     )
-    y1 = np.array([1, 1, 1, 1, 2, 2, 2, 2, 1, 1, 1, 1, 2, 2, 2, 2], dtype=int)
+    y = np.array([1, 1, 1, 1, 2, 2, 2, 2, 1, 1, 1, 1, 2, 2, 2, 2], dtype=int)
+    y = _enforce_estimator_tags_y(estimator_orig, y)
+    sw = np.ones(shape=len(y))
 
-    if kind == "ones":
-        X2 = X1
-        y2 = y1
-        sw2 = np.ones(shape=len(y1))
-        err_msg = (
-            f"For {name} sample_weight=None is not equivalent to sample_weight=ones"
-        )
-    elif kind == "zeros":
-        # Construct a dataset that is very different to (X, y) if weights
-        # are disregarded, but identical to (X, y) given weights.
-        X2 = np.vstack([X1, X1 + 1])
-        y2 = np.hstack([y1, 3 - y1])
-        sw2 = np.ones(shape=len(y1) * 2)
-        sw2[len(y1) :] = 0
-        X2, y2, sw2 = shuffle(X2, y2, sw2, random_state=0)
+    estimator1.fit(X, y=y, sample_weight=None)
+    estimator2.fit(X, y=y, sample_weight=sw)
 
-        err_msg = (
-            f"For {name}, a zero sample_weight is not equivalent to removing the sample"
-        )
-    else:  # pragma: no cover
-        raise ValueError
+    err_msg = f"For {name} sample_weight=None is not equivalent to sample_weight=ones"
+    for method in ["predict", "predict_proba", "decision_function", "transform"]:
+        if hasattr(estimator_orig, method):
+            X_pred1 = getattr(estimator1, method)(X)
+            X_pred2 = getattr(estimator2, method)(X)
+            assert_allclose_dense_sparse(X_pred1, X_pred2, err_msg=err_msg)
+
+
+@ignore_warnings(category=FutureWarning)
+def check_sample_weights_invariance(name, estimator_orig):
+    # check that setting sample_weight to zero / integer is equivalent
+    # to removing / repeating corresponding samples.
+    estimator_weighted = clone(estimator_orig)
+    estimator_repeated = clone(estimator_orig)
+    set_random_state(estimator_weighted, random_state=0)
+    set_random_state(estimator_repeated, random_state=0)
+
+    X = np.array(
+        [
+            [1, 4],
+            [1, 5],
+            [2, 4],
+            [2, 5],
+            [3, 1],
+            [3, 2],
+            [4, 1],
+            [4, 2],
+            [5, 4],
+            [5, 5],
+            [6, 4],
+            [6, 5],
+            [7, 1],
+            [7, 2],
+            [8, 1],
+            [8, 2],
+        ],
+        dtype=np.float64,
+    )
+    y = np.array([1, 1, 1, 1, 2, 2, 2, 2, 1, 1, 1, 1, 2, 2, 2, 2], dtype=int)
+
+    # Construct a dataset that is very different to (X, y) if weights
+    # are disregarded, but identical to (X, y) for nonzero weights.
+    X_weigthed = np.vstack([X, X + 1])
+    y_weighted = np.hstack([y, 3 - y])
+    # random integers and zero weights
+    rng = np.random.RandomState(0)
+    sw = rng.randint(1, 4, size=len(y_weighted))
+    sw[len(y) :] = 0
```

Comment:
That construction was designed specifically for 0 and 1 weights. Since we're extending the test for weights > 1, I think it could be simplified.
```suggestion
    X_weigthed = X  # we can give them this
    y_weighted = y  # name directly
    # random integers (including zero) weights
    rng = np.random.RandomState(0)
    sw = rng.randint(0, 4, size=len(y_weighted))
```